### PR TITLE
feat histogram

### PR DESCRIPTION
**Pull request by UchicagoZchen138**
Jira Ticket: [VADC-77](https://ctds-planx.atlassian.net/browse/VADC-77)

### New Features

- histogram endpoint

### CODE REVIEW

**Review by pieterlukasse — COMMENTED** [review]
hi @UchicagoZchen138, great PR! I’ve reviewed all the code and it looks very good. 

I've not reviewed the tests yet. Before I do that, I think it would be good to fix one issue:
- the histogram will have to reflect the _actual_ number of persons given the selected outcome and any selected covariates. This means that the we need to select these covariates and filter out all rows where any of the cells have a NA value.

File: controllers/cohortdata.go [hunk 1]
```diff
@@ -21,6 +21,56 @@ func NewCohortDataController(cohortDataModel models.CohortDataI) CohortDataContr
...
+	conceptIds, cohortPairs, err := utils.ParseConceptIdsAndDichotomousDefs(c)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "Error parsing request body for prefixed concept ids", "error": err.Error()})
+		c.Abort()
+		return
+	}
+
+	if len(conceptIds) == 0 {
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "must provide one or more concept ids for histogram endpoint"})
+		c.Abort()
```
Comment (pieterlukasse) [inline controllers/cohortdata.go:44]: this validation is not necessary. The code must also support the scenario where there are _no_ variables. If there are no variables, simply don't add the extra `join` statements to the query.

File: controllers/cohortdata.go [hunk 2]
```diff
@@ -21,6 +21,56 @@ func NewCohortDataController(cohortDataModel models.CohortDataI) CohortDataContr
...
+		return
+	}
+
+	// IMPORTANT ASSUMPTION: histogram concept_id is the last concept_id that we parse
+	histogramConceptId := conceptIds[len(conceptIds)-1]
```
Comment (pieterlukasse) [inline controllers/cohortdata.go:49]: I see... let's add this `histogramConceptId` as a separate concept id to the endpoint

File: models/cohortdata.go [hunk 3]
```diff
@@ -70,6 +70,29 @@ func (h CohortData) RetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPerso
...
 
+func (h CohortData) RetrieveHistogramDataBySourceIdAndCohortIdAndConceptIdsAndCohortPairs(sourceId int, cohortDefinitionId int, histogramConceptId int64, filterConceptIds []int64, filterCohortPairs []utils.CustomDichotomousVariableDef) ([]*PersonConceptAndValue, error) {
+	log.Printf(">> Using inner join impl. for large cohorts")
+	var dataSourceModel = new(Source)
+	omopDataSource := dataSourceModel.GetDataSource(sourceId, Omop)
+
+	resultsDataSource := dataSourceModel.GetDataSource(sourceId, Results)
+
+	// get the observations for the subjects and the concepts, to build up the data rows to return:
+	var histogramValueFieldName = "observation.value_as_number"
```
Comment (pieterlukasse) [inline models/cohortdata.go:81]: this one can just be hard coded in query, no need to make a variable of this.

File: models/cohortdata.go [hunk 4]
```diff
@@ -70,6 +70,29 @@ func (h CohortData) RetrieveDataBySourceIdAndCohortIdAndConceptIdsOrderedByPerso
...
+	var cohortData []*PersonConceptAndValue
+	query := omopDataSource.Db.Model(&Observation{}).
+		Select("distinct(observation.person_id), observation.observation_concept_id as concept_id, observation.value_as_string as concept_value_as_string, observation.value_as_number as concept_value_as_number, observation.value_as_concept_id as concept_value_as_concept_id").
+		Joins("INNER JOIN "+resultsDataSource.Schema+".cohort as cohort ON cohort.subject_id = observation.person_id").
```
Comment (pieterlukasse) [inline models/cohortdata.go:85]: let's keep only `observation.value_as_number` given this method is for numeric values only

File: models/cohortdata.go [hunk 5]
```diff
@@ -90,28 +113,8 @@ func (h CohortData) RetrieveCohortOverlapStats(sourceId int, caseCohortId int, c
...
-		cohortTableAlias1 := fmt.Sprintf("cohort_filter_1_%d", i)
-		cohortTableAlias2 := fmt.Sprintf("cohort_filter_2_%d", i)
-		unionAlias := "union_" + cohortTableAlias1 + "_" + cohortTableAlias2
-		log.Printf("Adding extra INNER JOIN on UNION with alias %s", unionAlias)
-		query = query.Joins("INNER JOIN (Select "+cohortTableAlias1+".subject_id,"+cohortTableAlias1+".cohort_definition_id FROM "+resultsDataSource.Schema+".cohort as "+cohortTableAlias1+
-			" UNION ALL Select "+cohortTableAlias2+".subject_id,"+cohortTableAlias2+".cohort_definition_id FROM "+resultsDataSource.Schema+".cohort as "+cohortTableAlias2+
-			") AS "+unionAlias+" ON "+unionAlias+".subject_id = observation.person_id").
-			Where(unionAlias+".cohort_definition_id in (?,?)", filterCohortPair.CohortId1, filterCohortPair.CohortId2)
-	}
+	query = QueryFilterByConceptIdsAndCohortPairsHelper(query, otherFilterConceptIds, filterCohortPairs, omopDataSource.Schema, resultsDataSource.Schema)
```
Comment (pieterlukasse) [inline models/cohortdata.go:116]: nice!

File: models/helper.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+func QueryFilterByConceptIdsAndCohortPairsHelper(query *gorm.DB, filterConceptIds []int64, filterCohortPairs []utils.CustomDichotomousVariableDef, omopSchemaName string, resultSchemaName string) *gorm.DB {
+	for i, filterConceptId := range filterConceptIds {
+		observationTableAlias := fmt.Sprintf("observation_filter_%d", i)
+		log.Printf("Adding extra INNER JOIN with alias %s", observationTableAlias)
+		query = query.Joins("INNER JOIN "+omopSchemaName+".observation as "+observationTableAlias+" ON "+observationTableAlias+".person_id = observation.person_id").
+			Where(observationTableAlias+".observation_concept_id = ?", filterConceptId).
+			Where("(" + observationTableAlias + ".value_as_string is not null or " + observationTableAlias + ".value_as_number is not null)") // TODO - improve performance by only filtering on type according to getConceptValueType()
+	}
+
+	for i, filterCohortPair := range filterCohortPairs {
```
Comment (pieterlukasse) [inline models/helper.go:20]: the original code had a comment above this loop. I think the comment is still useful. Maybe add it again (some variable names did change - so adjust that in the comment as well)?

**Review by pieterlukasse — COMMENTED** [review]
thanks @UchicagoZchen138 . I added some extra comments.

File: models/cohortdata.go (hunk 4 above)
Comment (UchicagoZchen138) [reply models/cohortdata.go:85]: can i keep the distinct(observation.person_id) and observation.value_as_concept_id as concept_value_as_concept_id for testing?

File: models/helper.go (hunk 6 above)
Comment (UchicagoZchen138) [reply models/helper.go:23]: fixed

File: models/cohortdata.go (hunk 4 above)
Comment (pieterlukasse) [reply models/cohortdata.go:84]: ah, I meant: from all the value_as_ fields, keep only the value_as_number :)

**Review by pieterlukasse — APPROVED** [review]
great work 👍

**Pull request merged**